File: Pascal/busca-binaria.pas
program buscabinaria;

const
    MAX = 30000;

type
    Tdado = integer;
    Tvetor = array[1..MAX]of integer;
	Tarq = text;
var 
  vetor : Tvetor;
  valor: Tdado;
	
procedure preencher(var vetor : Tvetor);
var
	i : integer;
begin
	for i:= 1 to MAX do
	begin	
		vetor[i] := i;
	end;
end;	

procedure buscaBinariarec(vetor:Tvetor; valor:Tdado; inicio:integer; fim:integer);
var 
	meio : integer;
begin
	meio := (inicio+fim) div 2;
	
    {
	writeln;	
	writeln('--------------------------------');
	writeln('Passo a Passo');
	writeln('valor	',vetor[meio]);
	writeln('posicao	',meio);
	writeln('--------------------------------');
	writeln;
    }
	
	if(valor = vetor[meio]) then
	begin
		writeln('Achou!!');
		//writeln('valor	',vetor[meio]);
		writeln('posicao	',meio);
	end else
		if(meio=inicio) then	write('error 404') else
		begin
		if((valor > vetor[meio]) and (meio <> inicio)) then
		begin
			//writeln('DIREITA');
			buscaBinariarec(vetor, valor, meio, MAX);
		end;
			
		if((valor < vetor[meio]) and (meio <> inicio)) then
		begin
			//writeln('ESQUERDA');
			buscaBinariarec(vetor, valor, inicio, meio);
		end;
	end;	
end;

begin
	valor := 21054;
	preencher(vetor);
	buscaBinariarec(vetor, valor, 0, MAX);
end.

File: src/ruby/BuscaBinaria.rb
# frozen_string_literal: true

def BuscaBinaria(valor, vetor, esquerda, direita)
  meio = ((esquerda + direita) / 2).floor

  return -1 unless esquerda <= direita

  if valor > vetor[meio]
    esquerda = meio + 1
    BuscaBinaria(valor, vetor, esquerda, direita)
  elsif valor < vetor[meio]
    direita = meio - 1
    BuscaBinaria(valor, vetor, esquerda, direita)
  else
    meio
  end
end

vetor = [0, 1, 3, 5, 6, 7, 8, 9, 10, 11, 12]
print BuscaBinaria(12, vetor, 0, vetor.length)
